File: extension/core/src/main/java/io/holunda/camunda/bpm/data/reader/RuntimeServiceVariableReader.java
package io.holunda.camunda.bpm.data.reader;

import io.holunda.camunda.bpm.data.factory.VariableFactory;
import org.camunda.bpm.engine.RuntimeService;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.Optional;

/**
 * Allows reading multiple variable values from {@link RuntimeService#getVariable(String, String)}.
 */
public class RuntimeServiceVariableReader implements VariableReader {

  private final RuntimeService runtimeService;
  private final String executionId;

  /**
   * Constructs a reader operating on execution.
   *
   * @param runtimeService runtime service to use.
   * @param executionId    execution id.
   */
  public RuntimeServiceVariableReader(RuntimeService runtimeService, String executionId) {
    this.runtimeService = runtimeService;
    this.executionId = executionId;
  }

  @NotNull
  @Override
  public <T> Optional<T> getOptional(VariableFactory<T> variableFactory) {
    return variableFactory.from(runtimeService, executionId).getOptional();
  }

  @NotNull
  @Override
  public <T> T get(VariableFactory<T> variableFactory) {
    return variableFactory.from(runtimeService, executionId).get();
  }

  @NotNull
  @Override
  public <T> T getLocal(VariableFactory<T> variableFactory) {
    return variableFactory.from(runtimeService, executionId).getLocal();
  }

  @NotNull
  @Override
  public <T> Optional<T> getLocalOptional(VariableFactory<T> variableFactory) {
    return variableFactory.from(runtimeService, executionId).getLocalOptional();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    RuntimeServiceVariableReader that = (RuntimeServiceVariableReader) o;

    if (!Objects.equals(runtimeService, that.runtimeService)) return false;
    return Objects.equals(executionId, that.executionId);
  }

  @Override
  public int hashCode() {
    int result = runtimeService != null ? runtimeService.hashCode() : 0;
    result = 31 * result + (executionId != null ? executionId.hashCode() : 0);
    return result;
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/reader/RuntimeServiceVariableReader.kt
package io.holunda.camunda.bpm.data.reader

import io.holunda.camunda.bpm.data.factory.VariableFactory
import org.camunda.bpm.engine.RuntimeService
import java.util.*

/**
 * Allows reading multiple variable values from [RuntimeService.getVariable].
 * @param runtimeService runtime service to use.
 * @param executionId    execution id.
 */
class RuntimeServiceVariableReader(private val runtimeService: RuntimeService, private val executionId: String) : VariableReader {
  override fun <T> getOptional(variableFactory: VariableFactory<T>): Optional<T> {
    return variableFactory.from(runtimeService, executionId).optional
  }

  override fun <T> get(variableFactory: VariableFactory<T>): T {
    return variableFactory.from(runtimeService, executionId).get()
  }

  override fun <T> getLocal(variableFactory: VariableFactory<T>): T {
    return variableFactory.from(runtimeService, executionId).local
  }

  override fun <T> getLocalOptional(variableFactory: VariableFactory<T>): Optional<T> {
    return variableFactory.from(runtimeService, executionId).localOptional
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val that = other as RuntimeServiceVariableReader
    return if (runtimeService != that.runtimeService) false else executionId == that.executionId
  }

  override fun hashCode(): Int {
    var result = runtimeService.hashCode()
    result = 31 * result + executionId.hashCode()
    return result
  }
}
